## Project Description

## Notebook Description

### Initialize packages and read in pickled data

In [9]:
% run __init__.py

In [11]:
cd ..

/home/jovyan


In [12]:
df_model = pd.read_pickle('data/df_model.p')

In [13]:
df_model.shape

(127052, 88)

### Try predicting our target (hit / no-hit) with a few models

**Remember our baseline metric:** If we guess hit for everything, we'd be right 32.6% of the time, or 67.4% of the time if we guessed no hit.

In [16]:
def run_benchmark(model, model_name, dataframe, target_col):
    target = dataframe[target_col]
    tmp_df = dataframe.drop(target_col, axis=1)
    X_train, X_test, y_train, y_test = train_test_split(tmp_df, target)
    model.fit(X_train, y_train)
    return {'train_score' : model.score(X_train, y_train), 
            'test_score' : model.score(X_test, y_test),
            'model_name' : model_name }

# credit to Joshua Cook

### K Neighbors Classifier

In [25]:
knn_output = run_benchmark(KNeighborsClassifier(),
                           'kneighbors',
                           df_model.drop(['player_id'], 
                                         axis=1), 
                           'hit_flag')

In [26]:
knn_output

{'model_name': 'kneighbors',
 'test_score': 0.79016465699083838,
 'train_score': 0.85389709200432373}

### Logistic Regression

In [27]:
log_reg_output = run_benchmark(LogisticRegression(), 
                               'logistic regression',
                               df_model.drop(['player_id'], 
                                         axis=1), 
                               'hit_flag')

In [28]:
log_reg_output

{'model_name': 'logistic regression',
 'test_score': 0.71992569971350318,
 'train_score': 0.71769039448414818}

### Decision Tree

In [29]:
dtree_output = run_benchmark(DecisionTreeClassifier(), 
                             'decision tree',
                             df_model.drop(['player_id'], 
                                         axis=1), 
                             'hit_flag')

In [30]:
dtree_output

{'model_name': 'decision tree',
 'test_score': 0.75660359537827027,
 'train_score': 1.0}

### Show benchmark models side-by-side

In [32]:
output = [
    knn_output,
    log_reg_output,
    dtree_output
]

pd.DataFrame(output)

model_name  test_score  train_score
0           kneighbors    0.790165     0.853897
1  logistic regression    0.719926     0.717690
2        decision tree    0.756604     1.000000

K Neighbors performed best, predicting with 79% accuracy on the test data. Compared to the ~67% by guessing no hit for everything, this is good but we can do better.

*Note: Despite 100% train score, decision tree is not overfit. The test score is pretty close to the other two models' test scores.*